darts/models/torch_forecasting_model.py

```diff
@@ -622,74 +631,70 @@ def _prepare_predict_tensors(self, n: int, input_series_dataset: TimeSeriesInfer
 
         Returns
         -------
-        torch.Tensor of shape `(len(input_series_dataset), self.input_chunk_length, target.width + covariates.width)`
-            Initial input to model which contains the input target sereis and the matching past covariates,
-            if available
-        Optional[torch.Tensor] of shape `(len(input_series_dataset), n, covariates.width)`
+        Sequence[TimeSeries]
+            Past target series which are going to be predicted into the future, and which will used as
+            input to the model.
+        Optional[Sequence[TimeSeries]]
+            Covariates required to predict horizons up to `self.output_chunk_length`, or `None` if the model
+            does not require covariates.
+        Optional[Sequence[TimeSeries]]
             Covariates required to predict horizons beyond `self.output_chunk_length`, or `None` if the model
             does require future covariates either because it does not require covariates at all or because
             `n <= self.output_chunk_length`.
     """
 
-        in_past_arr = []
+        tgt_past_arr = []
+        cov_past_arr = []
         cov_future_arr = []
 
-        for target_series, covariate_series in input_series_dataset:
+        covariates = [None] * len(series) if covariates is None else covariates
+        raise_if_not(len(series) == len(covariates),
+                     'The number of target series must be equal to the number of covariates.')
+
+        for target_series, covariate_series in zip(series, covariates):
```

Comment:
I'm a bit scared about this loop for large datasets.

Before, we had a "shallow" wrapping of the `Sequence[TimeSeries]` into a `SimpleInferenceDataset`, which means that the data doesn't have to be all loaded at once in memory (e.g. if the Sequence[TimeSeries] implements lazy loading). Now we are looping and building Lists which means everything will be held in memory.

I guess it can be fine (as it's for inference only - likely less often subject to big datasets - it wouldn't work for training), but it would be nice to find a way to do this split lazily to be really scalable in the 1% of cases where it's needed :D

Reply:
Absolutely agree, I removed the loop and moved the logic to the `SimpleInferenceDataset.__getitem__` method.